notebooks/02_model/mmlspark_lightgbm_criteo.ipynb

```diff
@@ -0,0 +1,399 @@
+    "One of the simplest ways of managing both features that have missing values as well as high cardinality is to use the hashing trick. The FeatureHasher transformer will pass integer values through and will hash categorical features into a sparse vector of lower dimensionality which can be used effectively by LightGBM.<br><br>\n",
+    "First the dataset is split randomly for training and testing and feature processing is applied to each dataset."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "raw_train, raw_test = spark_random_split(raw_data, ratio=0.8, seed=42)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "columns = [c for c in raw_data.columns if c != 'label']\n",
+    "feature_processor = FeatureHasher(inputCols=columns, outputCol='features')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "train = feature_processor.transform(raw_train)\n",
+    "test = feature_processor.transform(raw_test)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# save data to tables to finalize feature transformation\n",
+    "train.write.mode('overwrite').saveAsTable('train')\n",
+    "test.write.mode('overwrite').saveAsTable('test')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# load data from table\n",
+    "train = spark.table('train')\n",
+    "test = spark.table('test')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Model Training\n",
+    "In MML Spark the LightGBM implementation for binary classification is invoked using the LightGBMClassifier class and specifying the objective as 'binary'. In this instance the occurrence of positive labels is quite low, so setting the isUnbalance flag to true helps account for this imbalance.<br><br>\n",
+    "\n",
+    "### Hyper-parameters\n",
+    "Below are some of the key hyper-parameters \\[5\\] for training a LightGBM classifier on Spark\n",
+    "- numLeaves: the number of leaves in each tree\n",
+    "- numIterations: the number of iterations to apply boosting\n",
+    "- learningRate: the learning rate for training across trees\n",
+    "- featureFraction: the fraction of features used for training a tree\n",
+    "- earlyStoppingRound: round at which early stopping can be applied to avoid overfitting"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {
+    "tags": [
+     "parameters"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "NUM_LEAVES = 32\n",
+    "NUM_ITERATIONS = 10\n",
+    "LEARNING_RATE = 0.15\n",
+    "FEATURE_FRACTION = 0.8\n",
+    "EARLY_STOPPING_ROUND = 20"
```

Comment:
It makes more sense if NUM_ITERATIONS is greater than EARLY_STOPPING_ROUND.

Reply:
yes, we should increase iterations to 100, i just dropped it when it was taking a long time to train, but now it's not so bad

Reply:
this was changed by Andreas